server/lib/config/nginx.conf

```diff
@@ -105,7 +105,7 @@ http {
                 return 597;
             }
 
-            proxy_pass               http://127.0.0.1:8001;
+            proxy_pass               http://unix:/run/pbench-server.sock;
```

Comment:
Hopefully this doesn't pose any special NGINX issues when someone (e.g., our CI) attempts an API call before the socket exists? (Your test run apparently hasn't gotten that far, but I assume the answer will shortly be obvious!)

Reply:
I did run the jenkins/runlocal and it ran successfully so I am hoping it does not pose any issue in jenkins.

Reply:
Sigh! I guess there is some issue in the CI.

Reply:
I think all we need to do is have the container build tweak the Nginx systemd service file to declare that that Nginx depends on the Pbench Server.

Reply:
I'd prefer to have NGINX up early; I just don't want it to bomb because the server isn't ready. Going by the CI logs, NGINX is fine but gunicorn is unhappy and won't actually tell us why. I just see repeated

> pbench-server[72]: [2023-04-28 20:51:41 +0000] [72] [ERROR] Retrying in 1 second.

... with no indication I can see of *what* it's retrying or *why*... but the socket creation & setup seems like a reasonable guess. Does gunicorn require that the socket already exist, maybe??

> https://docs.gunicorn.org/en/stable/deploy.html

Shows a separate socket `systemd` unit that's required by the `gunicorn` unit, which seems to subtly suggest that it *does* need to be created separately. Though that doesn't explain why your `runlocal` worked... 😖

Reply:
My understanding is that Nginx won't be able to connect to the socket if it doesn't exist, by which I mean "have a listener connected to it".

My understanding of the `gunicorn.socket` file described in the Gunicorn doc is that it is a declaration/definition to `systemd` of the socket, so that the `gunicorn.service` file can declare a dependency on it.  The combination of these two has the effect that, when there is a connection attempt to the socket, `systemd` will start the service defined in the `gunicorn.service` file, with the expectation that it will listen on the socket.

However, in order to for `systemd` to detect the connection attempt, it probably creates the socket itself and adds some sort of listener to it (or some equivalent mechanism).  I don't know when Nginx will first reference the socket (i.e., I don't know if it sets up the connection "eagerly" or whether it defers it to the first HTTP request), but I believe that in our case Nginx is started as a `systemd` service; if it connects to the Pbench Server early, then that is (in general) a Good Thing™, but it does mean that the Pbench Server itself has to start pretty early, otherwise Nginx will get frustrated.

I don't think it matters whether the Pbench Server is started based solely on `After=network.target` or as a result of a `Requires=gunicorn.socket`, either way we ought to ensure that the proper sequencing is in place.  If `systemd` knows to start the Pbench Server based on the socket activity, then I assume that Nginx's attempt to connect to the socket will be held until Pbench can accept the connection.  Otherwise, I think we need to add a dependency to the Nginx service file to state that it depends on the Pbench Server service.  Failing either of those alternatives, we just need to kick off both services at about the same time, and Nginx just needs to be patient while the Pbench Server starts up.

Reply:
Right, as long as a socket exists with enough permissions for both the users to read and write at either end of the socket (in our case `pbench` and `nginx`), they should be able to communicate. I don't think the `systemd` needs to create a socket for them to communicate, it's one way to communicate but not the only way.  Gunicorn can create a socket on the fly when given a UNIX socket value for the `--bind` option with default permission of 777.